software/ds4_control.py

```diff
@@ -0,0 +1,224 @@
+from threading import Thread
+import motion
+from helper import map_range
+from pyPS4Controller.controller import Controller
+from enum import Enum
+
+
+class Axis(Enum):
+    """All possible movement axes"""
+    X = 1
+    Y = 2
+    ROT = 3
+    THROWER = 4
+
+
+class RobotDS4Backend(Controller):
+    def __init__(self, max_speed, analog_deadzone, default_thrower_speed, robot, **kwargs):
+        """Initializes RobotDS4Backend
+
+        Args:
+            max_speed (float): Max speed for robot in metres
+            analog_deadzone (int): Deadzone for analog stick
+            default_thrower_speed (int): Default speed to use for thrower
+            robot (OmniRobot): The robot to control
+        """
+        Controller.__init__(self, **kwargs)
+        # if there is no connection with the robot, open one
+        if robot == None:
+            self.robot = motion.OmniRobot()
+            self.robot.open()
+        else:
+            self.robot = robot
+        # parameters
+        self.max_speed = max_speed
+        self.analog_deadzone = analog_deadzone
+        # speed to use when we activate thrower
+        self.thrower_speed_param = default_thrower_speed
+        self.remote_controlled = False
+        # default is stopped, 0
+        self.x_speed = 0
+        self.y_speed = 0
+        self.rot_speed = 0
+        self.thrower_speed = 0
+        print("Controller added.")
+
+    def send_movement(self):
```

Comment:
It was from before we could easily get those values from main.py. I'll rewrite DS4 code a bit to make use of the new class in main.py.

It could cause issues with configurations done in main.py not applying to RC mode.

Reply:
Fixed.

Reply:
I would say that there are two main issues that can come from putting robot's logic into the controller's code. One is what you mentioned - if some configurations are done in one part of the code and not in the other, it could go out of sync. Second issue I see is the readability of the code (it might be a question of taste of course) - when I open a class dedicated for remote controlling, I would assume it hold the necessary logic to implement the remote controlling. And the main class of the robot would hold all the logic that 'makes the robit tick'. 

Moving is one of the main functionalities of the robot. Whether it gets it's inputs for moving from autonomus 'decisions' or from a controller is just a question of the situation, or let's say, a state :) Robot should be the main controller of the situations, as funky as it sounds - controller should just provide what is asked for. 

These comments are mainly as suggestions how it seems for me it would make the most sense. If it seems for you that the logic works better other ways, you are welcome to keep it that way.

Reply:
I'm now using all the variables I can from the main Robot class.

The actual moving still happens on the robot side, it's just that the autonomous side and controller side send the numbers.

The code makes sense to me, but if there's some way I could still improve it without rewriting the entire thing, let me know.